### PR TITLE
Converting HttpClientBuilders to interfaces

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

The hierarchy of client builders has recently been simplified. It is now
possible to simplify the structure even more by avoiding abstract
classes with intermediate implementations in the base abstract class for
particular case and using straightforward interfaces with default
implementations in the `http-netty` package.

Modifications:

- `MultiAddressHttpClientBuilder`, `PartitionedHttpClientBuilder`, and
  `SingleAddressHttpClientBuilder` are now interfaces,
- `DefaultMultiAddressUrlHttpClientBuilder`,
  `DefaultPartitionedHttpClientBuilder`, and
  `DefaultSingleAddressHttpClientBuilder` implement the above interfaces
  and the entire functionality is encapsulated within those final
  classes.

Result:

Cleaner API with interfaces and no intermediate abstract classes with
implementation scattered between the base and final classes.

### CODE REVIEW

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java [hunk 1]
```diff
@@ -68,16 +68,18 @@
     }
 
     @Override
-    public abstract PartitionedHttpClientBuilder<U, R> ioExecutor(IoExecutor ioExecutor);
+    PartitionedHttpClientBuilder<U, R> ioExecutor(IoExecutor ioExecutor);
 
     @Override
-    public abstract PartitionedHttpClientBuilder<U, R> executor(Executor executor);
+    default PartitionedHttpClientBuilder<U, R> executor(Executor executor) {
+        throw new UnsupportedOperationException("Setting Executor not yet supported by " + getClass().getSimpleName());
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java:75]: Same

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 2]
```diff
@@ -175,22 +157,23 @@
...
     }
 
     @Override
-    public abstract SingleAddressHttpClientBuilder<U, R> ioExecutor(IoExecutor ioExecutor);
+    SingleAddressHttpClientBuilder<U, R> ioExecutor(IoExecutor ioExecutor);
 
-    @Override
-    public abstract SingleAddressHttpClientBuilder<U, R> executor(Executor executor);
+    default SingleAddressHttpClientBuilder<U, R> executor(Executor executor) {
+        throw new UnsupportedOperationException("Setting Executor not yet supported by " + getClass().getSimpleName());
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:169]: Same

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 3]
```diff
@@ -72,7 +71,7 @@
      * data and log only network events.
      * @return {@code this}.
      */
-    public abstract SingleAddressHttpClientBuilder<U, R> enableWireLogging(String loggerName,
+    SingleAddressHttpClientBuilder<U, R> enableWireLogging(String loggerName,
                                                                                 LogLevel logLevel,
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:75]: Shifted indentation

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 4]
```diff
@@ -259,7 +242,7 @@
      * filtering.
      * @return {@code this}
      */
-    public SingleAddressHttpClientBuilder<U, R> appendClientFilter(Predicate<StreamingHttpRequest> predicate,
+    default SingleAddressHttpClientBuilder<U, R> appendClientFilter(Predicate<StreamingHttpRequest> predicate,
                                                                         StreamingHttpClientFilterFactory factory) {
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:246]: Shifted indentation

**Review by idelpivnitskiy — COMMENTED** [review]
LGTM after comments addressed.
If you don't mind, can we hold this PR as well until after #1792? Redirect should also be cherry-picked into 0.41. That will help to avoid conflicts in `MultiAddressHttpClientBuilder`.

**Review by bondolo — APPROVED** [review]
Please mark conversations as resolved once the PR incorporates the suggested changes. Leaving the conversations unresolved creates the impression that there are still additional changes to be made.  (Which can lead to the impression that review should be deferred until those changes are available).

**Comment by idelpivnitskiy** [comment]
Feel free to merge this PR after #1844 to close this task, don't wait for #1792. The conflicts will be minimal.

**Pull request merged**